Azure Monitor: Connect to backend with real data

sarahzinger opened the pull request:
**What this PR does / why we need it**:

Based off of https://github.com/grafana/grafana/pull/33879

Makes requests to azure resource graph for the metadata for the resource picker for our Logs (will also be used by Metrics)

**Which issue(s) this PR fixes**:

fixes https://github.com/grafana/grafana/issues/33700
<img width="832" alt="Screen Shot 2021-05-12 at 5 42 26 PM" src="https://user-images.githubusercontent.com/6620164/118048295-110dc900-b34a-11eb-9256-50351da0183c.png">

**Special notes for your reviewer**:

A note that because this PR adds a new route in plugin.json you will have to re-run `make run` for it to work properly. To preview the picker:
- create a panel (or edit an existing one)
- select "Azure Monitor"
- select "Logs" from service

TODO and some things that could be good to get feedback on (possibly in future prs):
- [ ] tests! I think I will focus mostly on testing the resource_picker_data file as I feel that is easy to test and also some of the most ugly complicated code
- [ ] I notice that both the "location" and the "type" columns look a little sad with real data in them, perhaps we can format them or map them to labels or something
- [ ] We could add a loading spinner if we think it's appropriate? I held off for now but we're easily set up for that.
- [ ] Any general thoughts on error handling would be welcome!
- [ ] Once we select our desired resource we then have to actually make the query with the selection. I haven't worked on that bit yet! I'm not sure if we want to do that in another PR perhaps? Maybe there's a way to feature gate this? Or we could just keep working on one mega branch.
- [ ] I need to verify that the where clause in the fetch for resources for a resource group is correct
- [ ] Some visual issues with data overflowing the modal now that we have real data.

sarahzinger committed:
Connect to backend with real data

andresmgot commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/resource_picker/resource_picker_data.ts:
```diff
@@ -0,0 +1,168 @@
+  '/resourcepickerdata/providers/Microsoft.ResourceGraph/resources?api-version=2020-04-01-preview';
+
+export default class ResourcePickerData {
+  private _proxyUrl: string;
+  private _formattedResourcePickerData: RowGroup;
```
not sure what's the policy here, but in the past I've used this linter rule  (https://eslint.org/docs/rules/no-underscore-dangle) that prevents people from using underscores in var names (since there is no need to do so).

sunker replied:
yeah we don't use underscores to mark it's private. see [frontend styleguide](https://github.com/grafana/grafana/blob/main/contribute/style-guides/frontend.md) (aka airbnb). 
that eslint rule would be nice to have!

andresmgot commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/NestedRows.tsx:
```diff
@@ -1,66 +1,98 @@
   const styles = useStyles2(getStyles);
 
+  const isSelected = !!selectedRows[row.id];
+  const isDisabled = Object.keys(selectedRows).length > 0 && !isSelected;
```
why is an entry disabled if it's not selected? (or am I reading this wrong?)

sarahzinger replied:
Yeah I didn't modify the behavior that was in [the original pr](https://github.com/grafana/grafana/pull/33879/files#diff-166cf4590c5fef3151bb77b3474c4c32ec73d045e7c0a95dcb22a82ff6fba59dR22) just reworked the data structure here. 

I think the idea is to demonstrate that you only can make one selection at a time so it forces you unselect your first choice before you can select another? I also found this a little clunky using it. I think I'd expect to be able to select something and it would auto unselect the other stuff. 

I can change that in this pr if we'd like!

andresmgot commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/NestedRows.tsx:
```diff
@@ -1,66 +1,98 @@
+      return;
+    }
+    setOpenStatus('loading');
+    await fetchNested(row);
```
should we skip the fetching if the row has been fetched before? (not sure if the content may change over time)

andresmgot commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/resource_picker/resource_picker_data.ts:
```diff
@@ -0,0 +1,168 @@
+      // TODO: verify that this is a good list (it's based off of what is found in the azure portal)
+      `resources 
+  | where resourceGroup == "${resourceGroup.name.toLowerCase()}"
+  | where type in ("microsoft.analysisservices/servers","microsoft.apimanagement/service","microsoft.network/applicationgateways","microsoft.insights/components","microsoft.web/hostingenvironments","microsoft.web/serverfarms","microsoft.web/sites","microsoft.automation/automationaccounts","microsoft.botservice/botservices","microsoft.appplatform/spring","microsoft.network/bastionhosts","microsoft ...
```
in which case would this list change? If we need to hardcode it I would move the different elements to arrays so if we need to add/remove/modify one, we don't need to change the whole line.

```
const types = [
  "microsoft.analysisservices/servers",
  ...
];
```

andresmgot commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/resource_picker/resource_picker_data.ts:
```diff
@@ -0,0 +1,168 @@
+  _formatAndCacheResourceData(rawData: any[]) {
+    // "produce" from immer takes in an object, allows you to edit a copy of that object and then returns a new object
+    this._formattedResourcePickerData = produce(this._formattedResourcePickerData, (draftState: any) => {
+      rawData.forEach((item: any) => {
```
it's probably worth it to define the type rather than using `any`

andresmgot reviewed (commented):
Hope you don't mind the questions/suggestions :) It's helpful to read this kind of PRs even though I don't have that much context

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/plugin.json:
```diff
@@ -122,6 +122,26 @@
       "headers": [{ "name": "x-ms-app", "content": "Grafana" }]
     },
+    {
+      "path": "resourcepickerdata",
```
I think we should call this azureresourcegraph

kostrse replied:
What if we add support for multi-verb routes? For example:
```
"path": "azure-management-api",
"method": "GET,POST"
```

sunker replied:
That would be nice! But as a part of [this](https://github.com/grafana/grafana/issues/33809), we'll have to refactor this and use a rest api endpoint in the plugin backend instead of using the plugin proxy. I guess this will be done for 8.1, so probably not worth putting any more effort into the pluginproxy at this point.

joshhunt replied:
Configuring Log Analytics to use different credentials complicates all this (even if we are deprecating it). There's three places we hit ARG for three different reasons. It's a messy venn diagram

 - Requesting client-side data for the resource picker UI for Log Analytics (with LA credentials, if configured)
 - Requesting resource picker UI for Metrics with the Metrics credentials
 - ARG datasource backend queries in PR #33293, with Metrics credentials.

That's why this route has its `resourcepickerdata` name to distinguish it from the endpoint that'll use Metrics credentials. 

This does need tidying up though!!!

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/resource_picker/resource_picker_data.ts:
```diff
@@ -0,0 +1,168 @@
+import { getBackendSrv } from '@grafana/runtime';
```
File names should be camelcase (snake case is for angular stuff)

sunker commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/resource_picker/resource_picker_data.ts:
```diff
@@ -0,0 +1,168 @@
+    return this._formattedResourcePickerData;
+  }
+
+  async _makeResourceGraphRequest(query: string, maxRetries = 1): Promise<any> {
```
this PR has some overlap with [this](https://github.com/grafana/grafana/pull/33293/files) which will be merged for beta2. This method (actually maybe everything in this file) should probably be moved to the [arg datasource file](https://github.com/grafana/grafana/blob/2112c3152e1b305ab12853fe7cc3c6e5a072abec/public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_resource_graph/azure_resource_graph_datasource.ts).

sarahzinger replied:
Yeah I'm not clear on if these requests should be hitting the same endpoints as those featured in that pr or not. 

The data I'm trying to fetch here seems very similar to me from what I would see in that PR so it felt to me like we could use the query method to get back this metadata. But then it sounded to me from Josh that Logs and Metrics can have different client ids that require a [different auth credential](https://github.com/grafana/grafana/issues/33765) and that a "datasource" is a unique concept different from these metadata type of requests? Perhaps we can discuss more on monday

joshhunt replied:
Theres is/maybe some overlap between these two things, but it's unclear to me right now what would be the same in practice _in the frontend_.

I think i would like to merge them both in and have a second pass cleaning up (perhaps related to the plugin proxy/auth refactor)

sunker reviewed (commented):
Haven't tested yet, but looks great Sarah. Regarding the things you wanted feedback on - let's talk tomorrow.

sarahzinger committed:
Fixes from code review.

sarahzinger committed:
WIP:Begin thinking about how to make queries with scope determined by the resource picker

joshhunt commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/NestedResourceTable.tsx:
```diff
@@ -5,21 +5,21 @@ import { useStyles2 } from '@grafana/ui';
+  selectedRows: RowGroup;
   noHeader?: boolean;
-  onRowToggleCollapse: (row: Row) => void;
+  fetchNested: (row: Row) => Promise<void>;
```
minor nit, but i think i would prefer a more... "api" like name that describes the event? Like requestNestedRows or even onExpandRow or something like that? idk. don't stress this one too much

joshhunt commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts:
```diff
@@ -0,0 +1,151 @@
+
+  _formatAndCacheResourceData(rawData: RawAzureResourceGraphDataItem[]) {
+    // "produce" from immer takes in an object, allows you to edit a copy of that object and then returns a new object
+    this.formattedResourcePickerData = produce(this.formattedResourcePickerData, (draftState: any) => {
```
hmmm... one thing that I've been trying to get away from is storing less state like this in these classes, and store/produce what we need in the react components themselves.

I think maintaining request caches in these classes is fine, but I think it would be simpler if these functions were a bit more... "functional" and just mapped over the data and returned the new format.

the `getResourcePickerDataWithNestedResourceData` function could then just be like `getResources`, returns just the list of resources, and then `fetchNested` back in the react component inserts them as the children. Thoughts?

sarahzinger replied:
Yeah I guess I liked the idea that we wouldn't make as many calls to the backend even if you switched between logs/metrics but you're probably right that it's weird to have this stuff hanging around hogging up memory even when the component is gone.

joshhunt commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts:
```diff
@@ -0,0 +1,151 @@
+    return this.formattedResourcePickerData;
+  }
+
+  async _makeResourceGraphRequest(query: string, maxRetries = 1): Promise<any> {
```
Would be nice if we could type this a little bit cleaner, without the any 😅

It looks like it should be something like `Promise<FetchResponse<{ data: RawAzureResourceGraphDataItem[] }>>` ?

Maybe we could make the response object (` { data: ... } `) its own interface?

joshhunt commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/supportedResources.ts:
```diff
@@ -0,0 +1,183 @@
+// TODO: should this list be different for logs vs metrics?
```
Good shout.

We'll need to see if we can get this list from Azure itself in the near future. I would rather not have to maintain this.

joshhunt reviewed (commented):
Gosh this is incredible! So exciting to see this coming along!!!

Just a few minor things we can discuss, otherwise I think this is looking really good. I'll push into the base branch the button/segment field that launches the modal.

sarahzinger committed:
More fixes post code review.

sarahzinger committed:
Fixes after rebasing

sarahzinger committed:
Remove outdated todo

sarahzinger commented:
@joshhunt ok I think I addressed all the comments, I still don't have tests which stinks but I am running out of time and figured I could pass this off for another review and come back to writing tests in a future pr if it still feels worth it. Tests are passing locally for me, but CI here has it marked as failing, I think because it doesn't know how to pull down your branch. 

If this seems suitable for merging, please feel free to do so no need to wait for me tomorrow morning!
